### PR TITLE
Rework overriding feature to allow for arbitrary depth of config and overrides

### PR DESCRIPTION
**Pull request by MarcDufresne**
Hello guys!

I've continued using Opset for more projects, and lately, I've started to feel some limitations around the structure of the config, especially around overrides with environment variables.

Here's an example:

```yaml
# default.conf
services:
  svc1:
    url: "http://sv1.app"
    api_key:
    router1:
      version: "v1"
```

Let's say I wanted to set the `api_key` of `services.svc1` it's not possible right now, neither via `local.yml` or environment variables, since we are limited to having only one section of depth. Level 2 must only contain values and can't _really_\* contain more sections.

*_Technically since we parse JSON it is possible to _define_ arbitrary structures, but they can only be overridden **completely** and not just in part._

What my change does is allow arbitrary structures to be defined in the `default.yml` and overridden via `local.yml` or env vars. It is demonstrated in the example below:

```yaml
# local.yml
services:
  svc1:
    api_key: "my-secret-api-key"
```

Applying the above to the default config from earlier would give you:

```yaml
services:
  svc1:
    url: "http://sv1.app"
    api_key: "my-secret-api-key"
    router1:
      version: "v1"
```

**NOTE**: If you did this in the current version the whole content of `services.svc1` would be replaced by the value from `local.yml`, so you would end up with just an `api_key` and would lose the rest.

And environment variables can now be used for any depth in the config, like so:

```shell script
MYAPP_SERVICES_SVC1_API_KEY="my-secret-api-key"
```

This change does not cause breaking changes. I've updated my apps to use this version and it's been seamless, and I also almost had no change to do to the unit tests. I've updated the version to `2.0.0-alpha0` anyway since this is a major rework of a critical feature.
Although you have the final say about the version number, just make sure to remove the `-alpha0` portion if you keep it at `2.0.0`.

Let me know what you think!

---

Before opening a PR, make sure that, if relevant:
- [x] Changes are covered by automated tests.
- [x] New / updated functions / methods have typehints.
- [x] READMEs have been updated.
- [x] The affected parties are / will be notified of incoming breaking changes.

### CODE REVIEW

**Comment by alexj-novisto** [comment]
👀

**Comment by FinchPowers** [comment]
+1! Albeit I don't understand why the build is not being triggered. I think it's because it needs to be pushed by someone in the EAI org. I will push a dummy commit to trigger it.

**Comment by FinchPowers** [comment]
Ok it's because it's a fork. I enabled CircleCI to build it. Can you push a dummy commit to trigger the build please? (Or force push a minor change to the existing one.)

File: Makefile [hunk 1]
```diff
@@ -21,7 +21,7 @@ lint: ## Lint code with flake8
 	poetry run mypy opset
 
 test: ## Run pytest test suite
-	poetry run pytest --cov=opset tests
+	poetry run pytest --cov-report term-missing --cov=opset tests
```
Comment (FinchPowers) [inline Makefile:24]: What is this for?

**Comment by MarcDufresne** [comment]
Cool, thanks!

File: opset/configurator.py [hunk 2]
```diff
@@ -1,17 +1,19 @@
 # configurator.py
-# Alexandre Jutras, 2018-11-19, Emilio Assuncao, 2019-01-17
+# Alexandre Jutras, 2018-11-19, Emilio Assuncao, 2019-01-17, Marc-Andre Dufresne, 2020-09-10
```
Comment (FinchPowers) [inline opset/configurator.py:2]: It should be a vertical list.

**Comment by MarcDufresne** [comment]
Haha, it fails because Coveralls can't push, since my fork doesn't have your token 🙃

File: Makefile (hunk 1 above)
Comment (FinchPowers) [reply Makefile:24]: Looking at your changes you should enable `--doctest-modules`. `_get_dict_item_from_path` docstring would get tested.

File: pyproject.toml [hunk 3]
```diff
@@ -1,6 +1,6 @@
 [tool.poetry]
 name = "opset"
-version = "1.5.0"
+version = "2.0.0-alpha.0"
```
Comment (FinchPowers) [inline pyproject.toml:3]: I think we can settle on 2.0.0.
However I don't know who will be able to publish it... :p

**Review by FinchPowers — APPROVED** [review]

**Comment by FinchPowers** [comment]
Test failure related to coveralls. I'm on it.

File: Makefile (hunk 1 above)
Comment (MarcDufresne) [reply Makefile:24]: I added this to be able to see the lines not covered without having to use coveralls, etc.

For the docstring it was mostly for the example, I also created a real unit test that tests the method directly

File: opset/configurator.py (hunk 2 above)
Comment (MarcDufresne) [reply opset/configurator.py:2]: I can change it if you wish

**Comment by FinchPowers** [comment]
Very weird, environment variables are missing from the build. Not sure if it can be related to the fork.

**Comment by MarcDufresne** [comment]
> Very weird, environment variables are missing from the build. Not sure if it can be related to the fork.

I think you need to enable an option like "Pass secrets to fork builds" (not sure about the name)

**Comment by FinchPowers** [comment]
Getting there...

**Comment by FinchPowers** [comment]
@MarcDufresne you have 2 unsigned commits

**Comment by MarcDufresne** [comment]
> @MarcDufresne you have 2 unsigned commits

If you squash merge won't it sign it with your account? I'm not set up to sign commits right now

**Comment by FinchPowers** [comment]
Negative: I can't squash merge either. The merge button is greyed out all together.

**Comment by MarcDufresne** [comment]
Ok let me try something

**Comment by MarcDufresne** [comment]
Ok what I did closed this PR... see this: https://github.com/ElementAI/opset/pull/19

**Pull request closed without merging**